Add RTEMS and QEMU RISCV support

peter-mitsis opened the pull request:
This set of commits does two things.
1. Add board support for both qemu_riscv32 and qemu_riscv64 for Zephyr.
2. Add support for the RTEMS RTOS (RISCV support).

RTEMS support on a RISCV board was tested on Zephyr's version of QEMU for RISCV, thereby providing a common base.

Note: The interrupt latency test is not yet working on RISCV+QEMU for either Zephyr or RTEMS. There are plans to continue that work, but it was thought desirable to share the work done thus far.

peter-mitsis committed:
zephyr: Add qemu_riscv support

Adds support for running the RTOS benchmark on both qemu_riscv64
and qemu_riscv32 with Zephyr. Although the neither board is expected
to provide accurate benchmark numbers, they should provide enough
infrastructure support to facilitate the addition of other
RISCV targets.

Signed-off-by: Peter Mitsis <peter.mitsis@intel.com>

peter-mitsis committed:
rtems: Create porting layer

Creates the porting layer for the RTEMS RTOS. This porting layer acts
as a translation between the generic RTOS routine names used in the
benchmark and the RTEMS RTOS.

Signed-off-by: Peter Mitsis <peter.mitsis@intel.com>

peter-mitsis committed:
rtos: Remove bench_offload_xxx() routines

The bench_offload_xxx() routines are not currently used in the
benchmark project.

Signed-off-by: Peter Mitsis <peter.mitsis@intel.com>

peter-mitsis committed:
rtos: add additional support macros

Not all RTOSes support arbitrary thread suspension and resume, nor
will all RTOSes use the existing main() as the entry point to the
benchmarking project. To accomodate for these differences, new
macros have been introduced in the various RTOS porting layers.

Signed-off-by: Peter Mitsis <peter.mitsis@intel.com>

peter-mitsis committed:
rtems: update RTEMS build specs

Adds infrastructure support to allow RTEMS to build the benchmark
project.

Signed-off-by: Peter Mitsis <peter.mitsis@intel.com>

peter-mitsis committed:
common: tweak thread tests

Tweaks the thread termination logic so that RTEMS does not run
out of thread resources.

Signed-off-by: Peter Mitsis <peter.mitsis@intel.com>

peter-mitsis committed:
rtems: Add README

Adds a README file that explains how to build and run the project
using RTEMS. This explanation assumes that the user has already
installed RTEMS.

Signed-off-by: Peter Mitsis <peter.mitsis@intel.com>

edersondisouza commented on src/zephyr/prj.qemu_riscv32.conf:
```diff
@@ -0,0 +1,34 @@
+CONFIG_TIMING_FUNCTIONS=y
+
+# Memory for pointer chasing
+CONFIG_HEAP_MEM_POOL_SIZE=4096
```
Do we still have pointer chasing?

peter-mitsis replied:
No we don't. I'll remove this.

edersondisouza commented on src/common/bench_all.c:
```diff
@@ -24,8 +24,10 @@ void bench_all(void *arg)
 	PRINTF("\n\r *** Done! ***\n\r");
```
On commit message s/artbitrary/arbitrary/

peter-mitsis replied:
Will do.

edersondisouza commented on h/bench_api.h:
```diff
@@ -12,6 +12,9 @@
 #ifdef FREERTOS
 #include "../freertos/bench_porting_layer_freertos.h"
 #endif /* FREERTOS */
+#ifdef RTEMS
```
I wonder if cmake can't help us here - there are two of these ifdef chains already...
Or maybe ensure bench_api has no need to leak anything from the rtos layer - why is it needed here?

peter-mitsis replied:
At the moment it is needed at least because of bench_time_t.

I agree that the code has an "ugly" feel to it. It is something that I would like to see cleaned up in the future.

edersondisouza commented on src/rtems/arch/riscv/core/arch_util.c:
```diff
@@ -0,0 +1,39 @@
+
+	sc = rtems_interrupt_handler_install(RISCV_INTERRUPT_VECTOR_TIMER,
+					"Clock",
+					RTEMS_INTERRUPT_REPLACE,
```
Shouldn't here be `RTEMS_INTERRUPT_UNIQUE` to avoid another handling doing stuff and skewing results?

peter-mitsis replied:
Can do.

edersondisouza commented on src/rtems/bench_porting_layer_rtems.h:
```diff
@@ -0,0 +1,42 @@
+#define RTOS_HAS_THREAD_SPAWN         1
+#define RTOS_HAS_THREAD_CREATE_START  0
+#define RTOS_HAS_SUSPEND_RESUME       0
+#define RTOS_HAS_MAIN_ENTRY_POINT     0
```
Will we be adding kconfig at some point? =P

peter-mitsis replied:
I think that is a future work item. I've left it this way for the moment so that each supported RTOS configures these options in the same way.

edersondisouza commented on src/rtems/timer/bench_riscv_machine_timer.c:
```diff
@@ -0,0 +1,123 @@
+/*
+ * Copyright (c) 2018 Intel Corporation
```
Is this date right? Anyway, is this the very same one file for zephyr?

peter-mitsis replied:
Good catch. Should be 2022.

peter-mitsis committed:
zephyr: remove CONFIG_HEAP_MEM_POOL_SIZE

The CONFIG_HEAP_MEM_POOL_SIZE parameter was only needed for
pointer chasing (which is no longer done).

Signed-off-by: Peter Mitsis <peter.mitsis@intel.com>

edersondisouza commented on src/rtems/bench_porting_layer_rtems.c:
```diff
@@ -0,0 +1,274 @@
+	(void) maximum_count;
+
+	rtems_name sem_name;
+	sem_name = rtems_build_name( 's', 'e', 'm', '4' );
```
Is it ok to have all semaphores having the same name?

peter-mitsis replied:
To my present knowledge, this is not a problem.

1. Unless I missed it, the documentation did not say that the name had to be unique.
2. There is a routine that converts a name to an ID (finding the 1st match).
3. If the name has to be unique, then that makes it little different from an ID. That in turn would likely create issues as various modules interact.

edersondisouza reviewed (approved):
LGTM, just a minor question.
